File: ios-web-port/HOneyCore/Sources/HOneyCore/CommunityV2/PostControls.swift
// Post controls on this iPhone (spec §35–§40; Web: lib/community-v2/
// local-store.ts + vault-client.ts + post-controls.ts). The active root and
// the wrapping root R live in the Keychain, sealed under a random device
// secret that never leaves the device; the Control Vault on Core holds the
// same roots as ciphertext under R; wrappers (recovery words · passkey PRF ·
// another device) only wrap R. Nothing here ever creates a second root
// while a server vault exists.

import Foundation

public struct LocalRoot: Sendable, Equatable {
    public var rootId: String
    public var secret: Data
    public var state: String
    public init(rootId: String, secret: Data, state: String) {
        self.rootId = rootId
        self.secret = secret
        self.state = state
    }
}

/// Roots unlocked on this device: the active root first.
public struct UnlockedRoots: Sendable, Equatable {
    public var vaultId: String
    public var r: Data
    public var roots: [LocalRoot]
    public var active: LocalRoot { roots[0] }
    public init(vaultId: String, r: Data, roots: [LocalRoot]) {
        self.vaultId = vaultId
        self.r = r
        self.roots = roots
    }
}

/// What the Keychain holds per account: R wrapped by the device secret, the
/// payload sealed under R, the last server revision and its wrappers.
public struct LocalVaultState: Codable, Sendable, Equatable {
    public var vaultId: String
    public var revision: Int
    public var wrappers: [VaultWrapper]
    public var rIv: String
    public var rWrapped: String
    public var payloadIv: String
    public var payloadCiphertext: String
}

public protocol PostControlStorage: Sendable {
    func loadState(account: String) throws -> LocalVaultState?
    func saveState(account: String, _ state: LocalVaultState) throws
    func clearState(account: String) throws
    /// A random 32-byte secret per account, created on first use; never exported.
    func deviceSecret(account: String) throws -> Data
    func clearDeviceSecret(account: String) throws
}

/// Keychain-backed storage (`SecretStore` is the Keychain abstraction the app injects).
public final class SecretPostControlStore: PostControlStorage, @unchecked Sendable {
    private let store: SecretStore
    private let prefix: String
    private let lock = NSLock()

    public init(store: SecretStore, prefix: String = "honey.v2") {
        self.store = store
        self.prefix = prefix
    }

    private func name(_ kind: String, _ account: String) -> String {
        "\(prefix).\(kind).\(AccountFiles.safeName(account))"
    }

    public func loadState(account: String) throws -> LocalVaultState? {
        lock.lock(); defer { lock.unlock() }
        guard let data = try store.read(name("vault", account)) else { return nil }
        return try? JSONDecoder().decode(LocalVaultState.self, from: data)
    }

    public func saveState(account: String, _ state: LocalVaultState) throws {
        lock.lock(); defer { lock.unlock() }
        try store.write(name("vault", account), try JSONEncoder().encode(state))
        guard let back = try store.read(name("vault", account)), (try? JSONDecoder().decode(LocalVaultState.self, from: back)) == state else {
            throw SecretStoreError.writeFailed
        }
    }

    public func clearState(account: String) throws {
        lock.lock(); defer { lock.unlock() }
        try store.delete(name("vault", account))
    }

    public func deviceSecret(account: String) throws -> Data {
        lock.lock(); defer { lock.unlock() }
        if let existing = try store.read(name("device", account)), existing.count == 32 { return existing }
        let secret = Data.random(32)
        try store.write(name("device", account), secret)
        guard try store.read(name("device", account)) == secret else { throw SecretStoreError.writeFailed }
        return secret
    }

    public func clearDeviceSecret(account: String) throws {
        lock.lock(); defer { lock.unlock() }
        try store.delete(name("device", account))
    }
}

/// The vault endpoints of Core the controls need (APIClient conforms).
public protocol VaultAPI: Sendable {
    func vault() async throws -> VaultRecord?
    func vaultPut(_ request: VaultPutRequest) async throws -> VaultPutResponse
    func vaultDelete() async throws
    func vaultPairingOffer(recipientPublicKey: String) async throws -> PairingOffer
    func vaultPairingRead(pairingId: String) async throws -> PairingOffer
    func vaultPairingDeliver(pairingId: String, enc: String, ciphertext: String) async throws
    func vaultPairingClaim(pairingId: String) async throws -> PairingDelivery?
}

public enum PostControlsStatus: Sendable, Equatable {
    /// No local root, no server vault → create silently at first share.
    case none
    /// Local root, no server vault yet → offer the encrypted backup.
    case localOnly(UnlockedRoots)
    /// A server vault exists but this device has no root → restore.
    case restoreNeeded(VaultRecord)
    case ready(UnlockedRoots, record: VaultRecord?, wrappers: [VaultWrapper])

    public var roots: UnlockedRoots? {
        switch self {
        case .localOnly(let r), .ready(let r, _, _): return r
        default: return nil
        }
    }
}

public enum PostControlsError: Error, Sendable, Equatable {
    case vaultExists
    case noLocalRoots
    case restoreNeeded
    case readbackFailed
    case wrongWords
    case pairingExpired
}

public actor PostControls {
    private let api: VaultAPI
    private let storage: PostControlStorage
    private let now: @Sendable () -> Int64

    public init(api: VaultAPI, storage: PostControlStorage, now: @escaping @Sendable () -> Int64 = { HOneyClock.now().epochMillis }) {
        self.api = api
        self.storage = storage
        self.now = now
    }

    // MARK: local

    /// The roots kept on this device, or nil.
    public func unlock(account: String) throws -> UnlockedRoots? {
        guard let state = try storage.loadState(account: account) else { return nil }
        let device = try storage.deviceSecret(account: account)
        let r = try VaultWrappers.unwrapWithDevice(deviceSecret: device, vaultId: state.vaultId, wrapped: WrappedR(iv: state.rIv, wrappedR: state.rWrapped))
        let payload = try ControlVault.open(r: r, vaultId: state.vaultId, revision: state.revision, iv: state.payloadIv, ciphertext: state.payloadCiphertext)
        return try Self.roots(vaultId: state.vaultId, r: r, payload: payload)
    }

    public func localState(account: String) throws -> LocalVaultState? {
        try storage.loadState(account: account)
    }

    private func save(account: String, roots: UnlockedRoots, payload: ControlVaultPayload, revision: Int, wrappers: [VaultWrapper]) throws {
        let device = try storage.deviceSecret(account: account)
        let wrapped = try VaultWrappers.wrapWithDevice(deviceSecret: device, vaultId: roots.vaultId, r: roots.r)
        let sealed = try ControlVault.seal(r: roots.r, vaultId: roots.vaultId, revision: revision, payload: payload)
        try storage.saveState(account: account, LocalVaultState(vaultId: roots.vaultId, revision: revision, wrappers: wrappers, rIv: wrapped.iv, rWrapped: wrapped.wrappedR, payloadIv: sealed.iv, payloadCiphertext: sealed.ciphertext))
    }

    static func roots(vaultId: String, r: Data, payload: ControlVaultPayload) throws -> UnlockedRoots {
        let active = try ControlVault.activeRoot(payload)
        var roots: [LocalRoot] = []
        for record in payload.roots {
            guard let secret = Base64URL.decode(record.secret) else { throw V2Error.vaultInvalid("root secret") }
            roots.append(LocalRoot(rootId: record.rootId, secret: secret, state: record.state))
        }
        let ordered = roots.filter { $0.rootId == active.rootId } + roots.filter { $0.rootId != active.rootId }
        return UnlockedRoots(vaultId: vaultId, r: r, roots: ordered)
    }

    /// The payload this device would upload (the vault content is derivable from what it holds).
    public func payload(for roots: UnlockedRoots, epochs: [SchoolEpoch]) -> ControlVaultPayload {
        let t = now()
        var payload = ControlVaultPayload(
            roots: roots.roots.map { ControlRootRecord(rootId: $0.rootId, secret: Base64URL.encode($0.secret), state: $0.state, createdAt: t) },
            activeRootId: roots.active.rootId,
            schoolEpochs: [],
            createdAt: t,
            updatedAt: t
        )
        for e in epochs { payload = ControlVault.withEpoch(payload, e, now: t) }
        return payload
    }

    /// Every school/year epoch known locally (plus the server's, when readable).
    public func epochs(account: String, roots: UnlockedRoots) async -> [SchoolEpoch] {
        var out: [SchoolEpoch] = []
        if let state = try? storage.loadState(account: account),
           let payload = try? ControlVault.open(r: roots.r, vaultId: state.vaultId, revision: state.revision, iv: state.payloadIv, ciphertext: state.payloadCiphertext) {
            out = payload.schoolEpochs
        }
        if let record = try? await api.vault(), let payload = try? ControlVault.open(r: roots.r, vaultId: record.vaultId, revision: record.revision, iv: record.iv, ciphertext: record.ciphertext) {
            for e in payload.schoolEpochs where !out.contains(e) { out.append(e) }
        }
        return out
    }

    // MARK: the decision table (spec §40.1)

    public func status(account: String) async throws -> PostControlsStatus {
        let local = try unlock(account: account)
        let record = try await api.vault()
        switch (local, record) {
        case (let local?, let record?):
            // Both: a newer server revision is merged into the local roots.
            if let state = try storage.loadState(account: account), record.revision > state.revision {
                let payload = try ControlVault.open(r: local.r, vaultId: record.vaultId, revision: record.revision, iv: record.iv, ciphertext: record.ciphertext)
                let roots = try Self.roots(vaultId: record.vaultId, r: local.r, payload: payload)
                try save(account: account, roots: roots, payload: payload, revision: record.revision, wrappers: record.wrappers)
                return .ready(roots, record: record, wrappers: record.wrappers)
            }
            return .ready(local, record: record, wrappers: record.wrappers)
        case (let local?, nil):
            return .localOnly(local)
        case (nil, let record?):
            return .restoreNeeded(record)
        case (nil, nil):
            return .none
        }
    }

    /// Create the initial root M1 and wrapping root R (no/no); nothing uploads until a wrapper exists.
    public func create(account: String, epoch: SchoolEpoch) async throws -> UnlockedRoots {
        if try await api.vault() != nil { throw PostControlsError.vaultExists }
        let m = Data.random(32)
        let r = Data.random(32)
        let t = now()
        let root = ControlVault.newRootRecord(secret: m, now: t)
        let payload = ControlVault.initialPayload(root: root, epochs: [epoch], now: t)
        let vaultId = "v_" + Base64URL.encode(.random(16))
        let roots = try Self.roots(vaultId: vaultId, r: r, payload: payload)
        try save(account: account, roots: roots, payload: payload, revision: 0, wrappers: [])
        return roots
    }

    /// The active roots for publishing; created silently when none exist anywhere.
    public func ensureRoots(account: String, epoch: SchoolEpoch) async throws -> UnlockedRoots {
        switch try await status(account: account) {
        case .ready(let roots, _, _), .localOnly(let roots): return roots
        case .none: return try await create(account: account, epoch: epoch)
        case .restoreNeeded: throw PostControlsError.restoreNeeded
        }
    }

    // MARK: upload with one merge retry (spec §40.2)

    @discardableResult
    public func upload(account: String, roots: UnlockedRoots, payload: ControlVaultPayload, wrappers: [VaultWrapper]) async throws -> VaultRecord {
        var base = try storage.loadState(account: account)?.revision ?? 0
        var current = payload
        var currentWrappers = wrappers
        for _ in 0..<2 {
            let revision = base + 1
            let sealed = try ControlVault.seal(r: roots.r, vaultId: roots.vaultId, revision: revision, payload: current)
            let result = try await api.vaultPut(VaultPutRequest(vaultId: roots.vaultId, baseRevision: base, iv: sealed.iv, ciphertext: sealed.ciphertext, wrappers: currentWrappers))
            switch result {
            case .ok(let rev, _):
                guard let record = try await api.vault(), record.revision == rev else { throw PostControlsError.readbackFailed }
                // Readback + decrypt-verify before anything reports success (spec §40.3 no.6).
                let verified = try ControlVault.open(r: roots.r, vaultId: record.vaultId, revision: record.revision, iv: record.iv, ciphertext: record.ciphertext)
                let merged = try Self.roots(vaultId: record.vaultId, r: roots.r, payload: verified)
                try save(account: account, roots: merged, payload: verified, revision: record.revision, wrappers: record.wrappers)
                return record
            case .conflict(let remoteRecord):
                guard remoteRecord.revision > 0, remoteRecord.vaultId == roots.vaultId else { throw V2Error.vaultConflict("a different vault exists for this account") }
                let remote = try ControlVault.open(r: roots.r, vaultId: remoteRecord.vaultId, revision: remoteRecord.revision, iv: remoteRecord.iv, ciphertext: remoteRecord.ciphertext)
                current = try ControlVault.merge(local: current, remote: remote, now: now())
                currentWrappers = ControlVault.mergeWrappers(local: currentWrappers, remote: remoteRecord.wrappers)
                base = remoteRecord.revision
            }
        }
        throw V2Error.vaultConflict("could not merge")
    }

    // MARK: recovery words (spec §37)

    /// Generates the 12 words, wraps R with them and uploads; the words are returned only after the readback decrypted.
    public func setupRecoveryWords(account: String, epochs: [SchoolEpoch]) async throws -> [String] {
        guard let roots = try unlock(account: account) else { throw PostControlsError.noLocalRoots }
        let secret = RecoveryWords.newSecret()
        let wrapper = try VaultWrappers.wrapWithPhrase(recoverySecret: secret, vaultId: roots.vaultId, r: roots.r, now: now())
        let existing = (try storage.loadState(account: account)?.wrappers ?? []).filter { if case .recoveryPhrase = $0 { return false } else { return true } }
        try await upload(account: account, roots: roots, payload: payload(for: roots, epochs: epochs), wrappers: existing + [.recoveryPhrase(wrapper)])
        return try RecoveryWords.words(from: secret)
    }

    /// Restore on a fresh device from the words (spec §37.3).
    public func restore(account: String, words: String) async throws -> UnlockedRoots {
        guard let secret = RecoveryWords.secret(from: words) else { throw PostControlsError.wrongWords }
        guard let record = try await api.vault() else { throw PostControlsError.noLocalRoots }
        for case .recoveryPhrase(let wrapper) in record.wrappers {
            if let r = try? VaultWrappers.unwrapWithPhrase(recoverySecret: secret, vaultId: record.vaultId, wrapper: wrapper) {
                return try restore(account: account, record: record, r: r)
            }
        }
        throw PostControlsError.wrongWords
    }

    /// With R recovered through any wrapper: decrypt the server vault and keep the roots locally.
    public func restore(account: String, record: VaultRecord, r: Data) throws -> UnlockedRoots {
        let payload = try ControlVault.open(r: r, vaultId: record.vaultId, revision: record.revision, iv: record.iv, ciphertext: record.ciphertext)
        let roots = try Self.roots(vaultId: record.vaultId, r: r, payload: payload)
        try save(account: account, roots: roots, payload: payload, revision: record.revision, wrappers: record.wrappers)
        return roots
    }

    // MARK: passkey PRF (spec §36) — the PRF output comes from the app layer (ASAuthorization)

    public func addPasskeyWrapper(account: String, prfOutput: Data, credentialId: String, label: String?, epochs: [SchoolEpoch]) async throws {
        guard let roots = try unlock(account: account) else { throw PostControlsError.noLocalRoots }
        let wrapper = try VaultWrappers.wrapWithPrf(prfOutput: prfOutput, vaultId: roots.vaultId, credentialId: credentialId, r: roots.r, now: now(), label: label)
        let existing = try storage.loadState(account: account)?.wrappers ?? []
        try await upload(account: account, roots: roots, payload: payload(for: roots, epochs: epochs), wrappers: existing + [.passkeyPrf(wrapper)])
    }

    public func restore(account: String, prfOutput: Data, credentialId: String) async throws -> UnlockedRoots {
        guard let record = try await api.vault() else { throw PostControlsError.noLocalRoots }
        for case .passkeyPrf(let wrapper) in record.wrappers where wrapper.credentialId == credentialId {
            let r = try VaultWrappers.unwrapWithPrf(prfOutput: prfOutput, vaultId: record.vaultId, wrapper: wrapper)
            return try restore(account: account, record: record, r: r)
        }
        throw PostControlsError.wrongWords
    }

    // MARK: another device (spec §38)

    /// New device: make an ephemeral key, register the offer, return the code + the private half to keep in memory.
    public func beginPairing(account: String) async throws -> (offer: PairingOffer, privateKey: String) {
        let pair = Pairing.newKeyPair()
        let offer = try await api.vaultPairingOffer(recipientPublicKey: pair.publicKey)
        return (offer, pair.privateKey)
    }

    /// Signed-in device: seal R to the offer's key.
    public func deliverPairing(account: String, pairingId: String) async throws {
        guard let roots = try unlock(account: account) else { throw PostControlsError.noLocalRoots }
        let offer = try await api.vaultPairingRead(pairingId: pairingId)
        let sealed = try Pairing.seal(recipientPublicKey: offer.recipientPublicKey, pairingId: offer.pairingId, r: roots.r)
        try await api.vaultPairingDeliver(pairingId: offer.pairingId, enc: sealed.enc, ciphertext: sealed.ciphertext)
    }

    /// New device: claim the delivery (nil while the other device has not answered) and restore.
    public func claimPairing(account: String, pairingId: String, privateKey: String) async throws -> UnlockedRoots? {
        guard let delivery = try await api.vaultPairingClaim(pairingId: pairingId) else { return nil }
        guard let record = try await api.vault() else { throw PostControlsError.noLocalRoots }
        let r = try Pairing.open(privateKey: privateKey, pairingId: delivery.pairingId, enc: delivery.enc, ciphertext: delivery.ciphertext)
        return try restore(account: account, record: record, r: r)
    }

    // MARK: rotation · erase · delete (spec §40.3, §40.4)

    public func rotate(account: String, epochs: [SchoolEpoch]) async throws -> UnlockedRoots {
        guard let roots = try unlock(account: account) else { throw PostControlsError.noLocalRoots }
        let rotated = ControlVault.rotated(payload(for: roots, epochs: epochs), newSecret: Data.random(32), now: now())
        let wrappers = try storage.loadState(account: account)?.wrappers ?? []
        let record = try await upload(account: account, roots: roots, payload: rotated, wrappers: wrappers)
        let verified = try ControlVault.open(r: roots.r, vaultId: record.vaultId, revision: record.revision, iv: record.iv, ciphertext: record.ciphertext)
        return try Self.roots(vaultId: record.vaultId, r: roots.r, payload: verified)
    }

    /// Remove the roots from this device only (the server vault stays; restore later through a wrapper).
    public func eraseLocal(account: String) throws {
        try storage.clearState(account: account)
        try storage.clearDeviceSecret(account: account)
    }

    /// Delete the server vault and the local roots (account deletion).
    public func deleteEverywhere(account: String) async throws {
        try await api.vaultDelete()
        try eraseLocal(account: account)
    }
}

File: apps/web/src/lib/community-v2/post-controls.ts
// Post Controls on the Web: the one object the pages talk to. It owns the
// current unlocked roots for the signed-in account, the school epochs, and
// the recovery methods (passkey PRF · recovery words · pairing). Every
// success it reports is a durable write that was read back and verified.

import {
  newPairingKeyPair,
  newRecoverySecret,
  openFromPairing,
  prfInput,
  sealForPairing,
  secretToWords,
  toBase64Url,
  unwrapWithPhrase,
  unwrapWithPrf,
  wordsToSecret,
  wrapWithPhrase,
  wrapWithPrf,
  type PasskeyPrfWrapper,
  type RecoveryPhraseWrapper,
  type VaultRecord,
  type VaultWrapper,
} from "@honey/shared/community-v2";
import { api, ApiError } from "../../api/client";
import { localPostControls, type UnlockedRoots } from "./local-store";
import { createPasskeyWithPrf, deviceLabel, getPrf } from "./passkey-prf";
import { vaultClient, type VaultStatus } from "./vault-client";

export interface Epoch {
  schoolId: string;
  academicYear: string;
}

async function currentEpoch(): Promise<Epoch> {
  const scope = await api.communityScope();
  return { schoolId: scope.schoolId, academicYear: scope.academicYear };
}

export const postControls = {
  status(account: string): Promise<VaultStatus> {
    return vaultClient.status(account);
  },

  /** Unlocked roots for publishing, or null when this device has none. */
  roots(account: string): Promise<UnlockedRoots | null> {
    return localPostControls.unlock(account);
  },

  /** First-time setup on a device with no root anywhere. */
  async create(account: string): Promise<UnlockedRoots> {
    const epoch = await currentEpoch();
    return vaultClient.createInitial(account, epoch);
  },

  async epochs(account: string, roots: UnlockedRoots): Promise<Epoch[]> {
    const epoch = await currentEpoch();
    const state = await localPostControls.load(account);
    const seen = new Map<string, Epoch>();
    seen.set(`${epoch.schoolId}\0${epoch.academicYear}`, epoch);
    // Epochs recorded in the vault payload cache, when present.
    if (state?.payloadCache) {
      try {
        const { openVault } = await import("@honey/shared/community-v2");
        const payload = await openVault(roots.r, { vaultId: roots.vaultId, ...state.payloadCache });
        for (const e of payload.schoolEpochs) seen.set(`${e.schoolId}\0${e.academicYear}`, e);
      } catch {
        /* cache unreadable: the current epoch is still right */
      }
    }
    return [...seen.values()];
  },

  // ---- recovery words -----------------------------------------------------

  /** Generate words + wrapper; nothing is saved until `commitRecoveryWords` (after the quiz). */
  async prepareRecoveryWords(roots: UnlockedRoots): Promise<{ words: string[]; wrapper: RecoveryPhraseWrapper }> {
    const secret = newRecoverySecret();
    const wrapper = await wrapWithPhrase(secret, roots.vaultId, roots.r, Date.now());
    return { words: secretToWords(secret), wrapper };
  },

  /** After the quiz: upload the vault with the phrase wrapper (replacing an older phrase wrapper). */
  async commitRecoveryWords(account: string, roots: UnlockedRoots, wrapper: RecoveryPhraseWrapper): Promise<VaultRecord> {
    const existing = (await localPostControls.load(account))?.wrappers ?? [];
    const wrappers: VaultWrapper[] = [...existing.filter((w) => w.type !== "recovery_phrase"), wrapper];
    const payload = vaultClient.payloadFor(roots, await this.epochs(account, roots));
    return vaultClient.upload(account, roots, payload, wrappers);
  },

  /** Restore from 12 words: the checksum is checked locally before any unwrap attempt. */
  async restoreWithWords(account: string, record: VaultRecord, words: string): Promise<UnlockedRoots | null> {
    const secret = wordsToSecret(words);
    if (!secret) return null;
    const wrapper = record.wrappers.find((w): w is RecoveryPhraseWrapper => w.type === "recovery_phrase");
    if (!wrapper) return null;
    try {
      const r = await unwrapWithPhrase(secret, record.vaultId, wrapper);
      return await vaultClient.restoreWithR(account, record, r);
    } catch {
      return null;
    }
  },

  // ---- passkey PRF ----------------------------------------------------------

  /** Create a passkey, verify with a fresh assertion that it unwraps R, then upload the wrapper. */
  async addPasskey(account: string, roots: UnlockedRoots, displayName: string): Promise<{ ok: true; record: VaultRecord } | { ok: false; reason: "unsupported" | "cancelled" | "verify_failed" }> {
    const input = prfInput(roots.vaultId);
    const created = await createPasskeyWithPrf(account, displayName, input);
    if (!created) return { ok: false, reason: "unsupported" };
    const wrapper = await wrapWithPrf(created.prfOutput, roots.vaultId, created.credentialId, roots.r, Date.now(), deviceLabel());
    // Readback: a fresh assertion must produce the same output and unwrap R.
    const again = await getPrf(created.credentialId, input);
    if (!again) return { ok: false, reason: "cancelled" };
    try {
      const r = await unwrapWithPrf(again.prfOutput, roots.vaultId, wrapper);
      if (toBase64Url(r) !== toBase64Url(roots.r)) return { ok: false, reason: "verify_failed" };
    } catch {
      return { ok: false, reason: "verify_failed" };
    }
    const existing = (await localPostControls.load(account))?.wrappers ?? [];
    const wrappers: VaultWrapper[] = [...existing.filter((w) => !(w.type === "passkey_prf" && w.credentialId === wrapper.credentialId)), wrapper];
    const payload = vaultClient.payloadFor(roots, await this.epochs(account, roots));
    const record = await vaultClient.upload(account, roots, payload, wrappers);
    return { ok: true, record };
  },

  /** Restore with any passkey wrapper the vault carries; null when none works here. */
  async restoreWithPasskey(account: string, record: VaultRecord): Promise<UnlockedRoots | null> {
    const wrappers = record.wrappers.filter((w): w is PasskeyPrfWrapper => w.type === "passkey_prf");
    if (wrappers.length === 0) return null;
    const input = prfInput(record.vaultId);
    // Let the platform pick among the vault's credentials (discoverable), then match the wrapper.
    const result = await getPrf(null, input);
    if (!result) return null;
    const wrapper = wrappers.find((w) => w.credentialId === result.credentialId) ?? null;
    if (!wrapper) return null;
    try {
      const r = await unwrapWithPrf(result.prfOutput, record.vaultId, wrapper);
      return await vaultClient.restoreWithR(account, record, r);
    } catch {
      return null;
    }
  },

  // ---- pairing --------------------------------------------------------------

  /** New device: announce an ephemeral key, show the code, wait for the signed-in device. */
  async beginPairing(): Promise<{ pairingId: string; expiresAt: number; privateKey: string }> {
    const kp = await newPairingKeyPair();
    const offer = await api.vaultPairingOffer(kp.publicKey);
    return { pairingId: offer.pairingId, expiresAt: offer.expiresAt, privateKey: kp.privateKey };
  },

  /** New device: poll until the ciphertext is there, then restore. */
  async completePairing(account: string, pairingId: string, privateKey: string): Promise<UnlockedRoots | null> {
    const delivery = await api.vaultPairingClaim(pairingId);
    if (!delivery) return null;
    const r = await openFromPairing(privateKey, pairingId, delivery.enc, delivery.ciphertext);
    const record = await api.vault();
    return vaultClient.restoreWithR(account, record, r);
  },

  /** Signed-in device: seal R to the code's key. */
  async approvePairing(roots: UnlockedRoots, pairingId: string): Promise<boolean> {
    let offer;
    try {
      offer = await api.vaultPairingRead(pairingId);
    } catch (err) {
      if (err instanceof ApiError && err.status === 404) return false;
      throw err;
    }
    const sealed = await sealForPairing(offer.recipientPublicKey, offer.pairingId, roots.r);
    await api.vaultPairingDeliver(offer.pairingId, sealed.enc, sealed.ciphertext);
    return true;
  },

  /** Same device, another HOney (Safari → installed app): a one-time link whose secret lives in the fragment. */
  async beginHandoff(roots: UnlockedRoots): Promise<string> {
    const kp = await newPairingKeyPair();
    // The relay binds the ciphertext to the pairing id, so: offer first, then seal and deliver.
    const offer = await api.vaultPairingOffer(kp.publicKey);
    const sealed = await sealForPairing(kp.publicKey, offer.pairingId, roots.r);
    await api.vaultPairingDeliver(offer.pairingId, sealed.enc, sealed.ciphertext);
    return `${location.origin}/settings/post-controls#handoff=${offer.pairingId}.${kp.privateKey}`;
  },

  /** Target side of the hand-off link. */
  async completeHandoff(account: string, fragment: string): Promise<UnlockedRoots | null> {
    const m = /^#?handoff=([A-Z0-9]{8})\.([A-Za-z0-9_-]+)$/.exec(fragment);
    if (!m) return null;
    return this.completePairing(account, m[1]!, m[2]!);
  },

  // ---- advanced ---------------------------------------------------------------

  async rotateRoot(account: string, roots: UnlockedRoots): Promise<UnlockedRoots> {
    const wrappers = (await localPostControls.load(account))?.wrappers ?? [];
    return vaultClient.rotate(account, roots, await this.epochs(account, roots), wrappers);
  },

  async eraseLocal(account: string): Promise<void> {
    await localPostControls.erase(account);
  },
};
